cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SparkBatchSourceFactory.java

```diff
@@ -61,32 +65,80 @@ public void addInput(String stageName, Input input) {
       addInput(stageName, ifpInput.getAlias(),
                new BasicInputFormatProvider(ifpInput.getInputFormatProvider().getInputFormatClassName(),
                                             ifpInput.getInputFormatProvider().getInputFormatConfiguration()));
+    } else if (input instanceof SQLEngineInput) {
+      addInput(stageName, input.getAlias(), (SQLEngineInput) input);
     }
   }
 
+  /**
+   * Add Dataset Input
+   * @param stageName stage name
+   * @param datasetName dataset name
+   * @param alias stage alias
+   * @param datasetArgs dataset arguments
+   * @param splits list of splits
+   */
   private void addInput(String stageName, String datasetName, String alias, Map<String, String> datasetArgs,
                         List<Split> splits) {
     duplicateAliasCheck(alias);
     datasetInfos.put(alias, new DatasetInfo(datasetName, datasetArgs, splits));
     addStageInput(stageName, alias);
   }
 
+  /**
+   * Add InputFormatProvider Input
+   * @param stageName stage name
+   * @param alias stage alias
+   * @param inputFormatProvider input format provider instance
+   */
   private void addInput(String stageName, String alias, InputFormatProvider inputFormatProvider) {
     duplicateAliasCheck(alias);
     inputFormatProviders.put(alias, inputFormatProvider);
     addStageInput(stageName, alias);
   }
 
+  /**
+   * Add SQL Engine Input
+   * @param stageName stage name
+   * @param alias stage alias
+   * @param sqlEngineInput input
+   */
+  private void addInput(String stageName, String alias,
+                         SQLEngineInput sqlEngineInput) {
+    if (sqlInputs.containsKey(stageName)) {
+      throw new IllegalArgumentException("Input already configured: " + alias);
+    }
+    sqlInputs.put(alias, sqlEngineInput);
+    addStageInput(stageName, alias);
+
+    // Add fallback input as well
+    addInput(stageName, sqlEngineInput.getFallbackInput());
```

Comment:
We don't need the fallback input anymore, do we? 

Since a SQL input will fail if it cannot read from SQL source, this is no longer required.